test/end2end_test.go

```diff
@@ -2432,6 +2432,102 @@ func testHealthCheckOff(t *testing.T, e env) {
 	}
 }
 
+func TestHealthWatch(t *testing.T) {
+	defer leakcheck.Check(t)
+	for _, e := range listTestEnv() {
+		testHealthWatch(t, e)
+	}
+}
+
+func testHealthWatch(t *testing.T, e env) {
+	te := newTest(t, e)
+	hs := health.NewServer()
+
+	service0 := ""
+	service1 := "grpc.health.v1.Health1"
+	service2 := "grpc.health.v1.Health2"
+	te.healthServer = hs
+
+	hs.SetServingStatus(service1, healthpb.HealthCheckResponse_SERVING)
+
+	te.startServer(&testServer{security: e.security})
+	defer te.tearDown()
+
+	cc := te.clientConn()
+	ctx10, cancel10 := context.WithCancel(context.Background())
+	defer cancel10()
+	ctx11, cancel11 := context.WithCancel(context.Background())
+	defer cancel11()
+	ctx12, cancel12 := context.WithCancel(context.Background())
+	defer cancel12()
+	ctx21, cancel21 := context.WithCancel(context.Background())
+	defer cancel21()
+
+	hc11 := healthgrpc.NewHealthClient(cc)
+	req0 := &healthpb.HealthCheckRequest{
+		Service: service0,
+	}
+	hc12 := healthgrpc.NewHealthClient(cc)
+	req1 := &healthpb.HealthCheckRequest{
+		Service: service1,
+	}
+
+	hc21 := healthgrpc.NewHealthClient(cc)
+	req2 := &healthpb.HealthCheckRequest{
+		Service: service2,
+	}
+
+	stream10, err := hc11.Watch(ctx10, req0)
+	if err != nil {
+		t.Fatalf("error: %v", err)
+	}
+
+	healthWatchChecker(t, stream10, healthpb.HealthCheckResponse_SERVING)
+
+	stream11, err := hc11.Watch(ctx11, req1)
+	if err != nil {
+		t.Fatalf("error: %v", err)
+	}
+
+	healthWatchChecker(t, stream11, healthpb.HealthCheckResponse_SERVING)
+
+	hs.SetServingStatus(service0, healthpb.HealthCheckResponse_NOT_SERVING)
+	healthWatchChecker(t, stream10, healthpb.HealthCheckResponse_NOT_SERVING)
+
+	stream21, err := hc21.Watch(ctx21, req2)
+	if err != nil {
+		t.Fatalf("error: %v", err)
+	}
+
+	healthWatchChecker(t, stream21, healthpb.HealthCheckResponse_SERVICE_UNKNOWN)
+	hs.SetServingStatus(service2, healthpb.HealthCheckResponse_NOT_SERVING)
+	healthWatchChecker(t, stream21, healthpb.HealthCheckResponse_NOT_SERVING)
+
+	stream12, err := hc12.Watch(ctx12, req1)
+	if err != nil {
+		t.Fatalf("error: %v", err)
+	}
+
+	healthWatchChecker(t, stream12, healthpb.HealthCheckResponse_SERVING)
+
+	hs.SetServingStatus(service2, healthpb.HealthCheckResponse_SERVING)
+	hs.SetServingStatus(service1, healthpb.HealthCheckResponse_NOT_SERVING)
+
+	healthWatchChecker(t, stream11, healthpb.HealthCheckResponse_NOT_SERVING)
```

Comment:
Also do
```go
healthWatchChecker(t, stream12, healthpb.HealthCheckResponse_NOT_SERVING)
```
?

This is to check the update to `service1` would affect both streams watching `service1`.